### PR TITLE
(WIP) many changes to investigate indeterminstic compute

### PR DESCRIPTION
**Pull request by liqunfu**
**Description**: Describe your changes.
experimental code to investigate indeterministic compute

**Motivation and Context**

### CODE REVIEW

File: onnxruntime/core/providers/cuda/gpu_data_transfer.cc [hunk 1]
```diff
@@ -60,4 +64,78 @@ common::Status GPUDataTransfer::CopyTensor(const Tensor& src, Tensor& dst, int e
   return Status::OK();
 }
 
+void PrintTensor(const Tensor& tensor, const std::string& name, bool summary)
```
Comment (KeDengMS) [inline onnxruntime/core/providers/cuda/gpu_data_transfer.cc:67]: >PrintTensor [](start = 5, length = 11)

This is nice to have more debuggability. Moreover, should we consider a macro to turn it off when not needed, to save some binary size?

File: onnxruntime/core/providers/cuda/gpu_data_transfer.cc [hunk 2]
```diff
@@ -60,4 +64,78 @@ common::Status GPUDataTransfer::CopyTensor(const Tensor& src, Tensor& dst, int e
...
+      }
+      std::cout << name << ": " << std::setprecision(std::numeric_limits<float>::digits10 + 3) << sum << " sq: "<< sum_sq << std::endl;
+    } else {
+      // save the data
+      std::string filename = name;
+      std::replace(filename.begin(), filename.end(), ':', '-');
+      std::replace(filename.begin(), filename.end(), ' ', '-');
+      std::replace(filename.begin(), filename.end(), '/', '-');
+
+      std::ofstream wf(std::string("/bert_ort/liqun/test_out/") + filename + ".dat", std::ios::out | std::ios::binary);
```
Comment (KeDengMS) [inline onnxruntime/core/providers/cuda/gpu_data_transfer.cc:96]: >/bert_ort/liqun/test_out [](start = 36, length = 24)

hardcoded?

File: onnxruntime/core/providers/cuda/gpu_data_transfer.cc [hunk 3]
```diff
@@ -60,4 +64,78 @@ common::Status GPUDataTransfer::CopyTensor(const Tensor& src, Tensor& dst, int e
...
+        }
+        for (int count = 0; count < total; count++) {
+          float sum = 0;
+          for (int i = 0; i < stride; i++) {
+            sum += tensor_cpu[count * stride + i];
+          }
+          std::cout << name << ": " << std::setprecision(std::numeric_limits<float>::digits10 + 3) << sum << std::endl;
+        }
+      }
+    }
```
Comment (KeDengMS) [inline onnxruntime/core/providers/cuda/gpu_data_transfer.cc:121]: could this be templatized so we may support fp16 in future?

File: onnxruntime/core/providers/cuda/math/binary_elementwise_ops.cc [hunk 4]
```diff
@@ -404,6 +404,12 @@ Status VariadicInputBase<T, CudaT>::ComputeMethod(OpKernelContext* context, Impl
           prepare.fdm_C,
           reinterpret_cast<CudaT*>(prepare.output_tensor->template MutableData<T>()),
           prepare.output_tensor->Shape().Size());
+          // if(context->NodeName() == "AccumulateGrad_63_grad" || context->NodeName() == "AccumulateGrad_66_grad") {
+          //   for (int index = 0; index < input_count; index++) {
+          //     PrintTensor(*context->Input<Tensor>(index), "Sum: " + context->NodeInputName(index), true);
+          //   }
+          //   PrintTensor(*output_tensor, "Sum: " + context->NodeOutputName(0), true);
+          // }
```
Comment (KeDengMS) [inline onnxruntime/core/providers/cuda/math/binary_elementwise_ops.cc:412]: not needed?

File: onnxruntime/core/providers/cuda/reduction/reduction_ops.cc [hunk 5]
```diff
@@ -116,7 +116,7 @@ Status ReduceKernel<allow_multi_axes>::ReduceKernelShared(
   // Block of fast matrix row reduction.
   const auto stride = input_shape[input_shape.NumDimensions() - 1];
   const auto reduction_size = input_shape.Size() / stride;
-  if (fast_reduction_ && reduction_size <= std::numeric_limits<int>::max() && stride <= std::numeric_limits<int>::max() &&
+  if (false && fast_reduction_ && reduction_size <= std::numeric_limits<int>::max() && stride <= std::numeric_limits<int>::max() &&
```
Comment (KeDengMS) [inline onnxruntime/core/providers/cuda/reduction/reduction_ops.cc:119]: >false [](start = 6, length = 5)

Could we add a global flag to control this?

File: orttraining/orttraining/core/framework/gradient_graph_builder.cc [hunk 6]
```diff
@@ -221,7 +228,22 @@ Status GradientGraphBuilder::Build() {
...
+    std::cout << (*it)->Name() << std::endl;
+
+  std::cout << "--------------" << "GetNodesInTopologicalOrder" << std::endl;
+  const std::vector<NodeIndex>& ordered_indices = graph_->GetNodesInTopologicalOrder();
+  for (size_t i = 0; i < ordered_indices.size(); i++) {
+    const NodeIndex& node_index = ordered_indices[i];
+    auto& node = *graph_->GetNode(node_index);
+    std::cout << node.Name() << std::endl;
+  }
+
```
Comment (KeDengMS) [inline orttraining/orttraining/core/framework/gradient_graph_builder.cc:245]: consider to put this under debug macro

File: orttraining/orttraining/core/framework/gradient_graph_builder.h [hunk 7]
```diff
@@ -22,7 +23,13 @@
...
 namespace training {
 
-typedef std::unordered_set<const Node*> NodeSet;
+// typedef std::unordered_set<const Node*> NodeUnorderedSet;
+struct node_compare {
+    bool operator() (const Node* lhs, const Node* rhs) const {
+        return lhs->Name() < rhs->Name();
+    }
+};
+typedef std::set<const Node*, node_compare> NodeSet;
```
Comment (KeDengMS) [inline orttraining/orttraining/core/framework/gradient_graph_builder.h:32]: >NodeSet [](start = 44, length = 7)

please add a comment about why this affects gradient graph. Plus, there are some other undered_set/unordered_map in here, like x/y_node_arg_names. Should that change to set too?

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 8]
```diff
@@ -226,6 +226,8 @@ void addObjectMethodsForTraining(py::module& m) {
         std::vector<std::string> provider_types = {};
         InitializeSession(sess, provider_types);
 
+        const std::string model_uri("/bert_ort/liqun/test_out/hf_bert_backprop.onnx");
+        sess->Save(model_uri, onnxruntime::training::TrainingSession::SaveOption::NO_RELOAD);
```
Comment (KeDengMS) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:230]: remove or put under debug macro?

File: orttraining/orttraining/test/python/orttraining_run_glue.py [hunk 9]
```diff
@@ -128,6 +128,8 @@ def run_glue(self, model_name, task_name, fp16):
             model_args.tokenizer_name if model_args.tokenizer_name else model_args.model_name_or_path,
             cache_dir=model_args.cache_dir,
         )
+
+        # config.num_hidden_layers = 1
```
Comment (KeDengMS) [inline orttraining/orttraining/test/python/orttraining_run_glue.py:132]: >      # config.num_hidden_layers = 1 [](start = 2, length = 36)

remove if not needed

**Comment by tianleiwu** [comment]
You can build binary with --cmake_extra_defines onnxruntime_DEBUG_NODE_INPUTS_OUTPUTS=2, and then tensors will be print to stdout during inference. That is supported right now. Maybe you can try it in training as well.

**Pull request closed without merging**